File: lib/scheduler.py
"""Планировщик внутри сервера: время выбирается в интерфейсе, cron не нужен.

Пропущенные из-за выключенной машины запуски догоняются, а сработавшие слоты
помечаются в data/schedule-state.json, чтобы не запускаться дважды за день.
"""

from __future__ import annotations

import datetime
import threading
import time
from typing import Any, Dict, Optional

from . import store
from .settings import get_settings

TICK_SECONDS = 30
_started = False


def _zone(name: str):
    try:
        from zoneinfo import ZoneInfo

        return ZoneInfo(name)
    except Exception:
        return datetime.timezone.utc


def now_in(timezone: str) -> Dict[str, Any]:
    now = datetime.datetime.now(_zone(timezone))
    return {"hhmm": now.strftime("%H:%M"), "date": now.strftime("%Y-%m-%d"), "minutes": now.hour * 60 + now.minute}


def _to_minutes(hhmm: str) -> int:
    try:
        hours, minutes = hhmm.split(":")
        return int(hours) * 60 + int(minutes)
    except Exception:
        return 0


def next_run(times, timezone: str) -> Optional[Dict[str, Any]]:
    if not times:
        return None
    now = now_in(timezone)
    ordered = sorted(times)
    upcoming = next((t for t in ordered if _to_minutes(t) > now["minutes"]), None)
    if upcoming:
        return {"at": upcoming, "inMinutes": _to_minutes(upcoming) - now["minutes"], "today": True}
    return {"at": ordered[0], "inMinutes": 1440 - now["minutes"] + _to_minutes(ordered[0]), "today": False}


def _tick() -> None:
    from .agent import run_agent

    schedule = get_settings()["schedule"]
    if not schedule["enabled"] or not schedule["times"]:
        return

    now = now_in(schedule["timezone"])
    state = store.read_json("schedule-state.json", {"fired": {}, "lastRunAt": None}) or {"fired": {}, "lastRunAt": None}
    fired = state.get("fired") or {}

    for slot in schedule["times"]:
        if fired.get(slot) == now["date"]:
            continue
        due = _to_minutes(slot) <= now["minutes"] if schedule["catchUp"] else slot == now["hhmm"]
        if not due:
            continue

        # помечаем слот до запуска: разбор длится минутами, а тик идёт каждые 30 секунд
        fired[slot] = now["date"]
        state["fired"] = fired
        state["lastRunAt"] = datetime.datetime.now(datetime.timezone.utc).isoformat()
        store.write_json("schedule-state.json", state)

        try:
            run = run_agent("cron")
            print(
                "[планировщик] %s %s: портфель $%d, сигналов %d, telegram: %s"
                % (
                    slot,
                    schedule["timezone"],
                    round(run["totalValueUsd"]),
                    len(run["alerts"]),
                    "отправлено" if run["telegram"]["sent"] else run["telegram"].get("error"),
                ),
                flush=True,
            )
        except Exception as e:
            print("[планировщик] %s: разбор упал — %s" % (slot, e), flush=True)


def start_scheduler() -> None:
    global _started
    if _started:
        return
    _started = True

    def loop():
        time.sleep(5)  # первый тик сразу, чтобы догнать пропущенное после перезапуска
        while True:
            try:
                _tick()
            except Exception as e:
                print("[планировщик] ошибка тика: %s" % e, flush=True)
            time.sleep(TICK_SECONDS)

    threading.Thread(target=loop, daemon=True).start()
    print("[планировщик] запущен, проверка каждые %d секунд" % TICK_SECONDS, flush=True)

File: src/lib/scheduler.ts
import { runAgent } from "./agent";
import { readJson, writeJson } from "./store";
import { getSettings } from "./telegram";

type ScheduleState = {
  /** "09:00" -> "2026-08-20", чтобы не запускать слот дважды за день */
  fired: Record<string, string>;
  lastRunAt: string | null;
};

const TICK_MS = 30_000;
let started = false;

/** Время и дата в часовом поясе пользователя, а не сервера. */
export function nowIn(timeZone: string): { hhmm: string; date: string; minutes: number } {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone,
    hour12: false,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
  }).formatToParts(new Date());
  const get = (t: string) => parts.find((p) => p.type === t)?.value ?? "00";
  const hour = get("hour") === "24" ? "00" : get("hour");
  return {
    hhmm: `${hour}:${get("minute")}`,
    date: `${get("year")}-${get("month")}-${get("day")}`,
    minutes: Number(hour) * 60 + Number(get("minute")),
  };
}

const toMinutes = (hhmm: string) => {
  const [h, m] = hhmm.split(":").map(Number);
  return (h || 0) * 60 + (m || 0);
};

/** Следующий запуск и сколько до него — для панели расписания. */
export function nextRun(times: string[], timeZone: string): { at: string; inMinutes: number; today: boolean } | null {
  if (!times.length) return null;
  const now = nowIn(timeZone);
  const sorted = [...times].sort();
  const upcoming = sorted.find((t) => toMinutes(t) > now.minutes);
  if (upcoming) return { at: upcoming, inMinutes: toMinutes(upcoming) - now.minutes, today: true };
  return { at: sorted[0], inMinutes: 1440 - now.minutes + toMinutes(sorted[0]), today: false };
}

async function tick(): Promise<void> {
  const settings = await getSettings();
  const { schedule } = settings;
  if (!schedule.enabled || !schedule.times.length) return;

  const now = nowIn(schedule.timezone);
  const state = await readJson<ScheduleState>("schedule-state.json", { fired: {}, lastRunAt: null });

  for (const time of schedule.times) {
    if (state.fired[time] === now.date) continue;

    const due = schedule.catchUp ? toMinutes(time) <= now.minutes : time === now.hhmm;
    if (!due) continue;

    // помечаем слот до запуска: разбор длится десятки секунд, а тик идёт каждые 30
    state.fired[time] = now.date;
    state.lastRunAt = new Date().toISOString();
    await writeJson("schedule-state.json", state);

    try {
      const run = await runAgent("cron");
      console.log(
        `[scheduler] ${time} ${schedule.timezone}: портфель $${Math.round(run.totalValueUsd)}, сигналов ${run.alerts.length}, telegram: ${
          run.telegram.sent ? "отправлено" : run.telegram.error
        }`,
      );
    } catch (err) {
      console.error(`[scheduler] ${time}: разбор упал —`, err instanceof Error ? err.message : err);
    }
  }
}

/**
 * Планировщик внутри приложения: пользователь выбирает время в интерфейсе,
 * crontab не нужен. Пропущенные из-за выключенной машины запуски догоняются.
 */
export function startScheduler(): void {
  if (started) return;
  started = true;
  console.log("[scheduler] запущен, проверка каждые 30 секунд");
  setInterval(() => {
    tick().catch((err) => console.error("[scheduler] tick error:", err));
  }, TICK_MS).unref?.();
  // первый тик сразу — чтобы догнать пропущенное после перезапуска
  setTimeout(() => tick().catch(() => {}), 5_000);
}
